src/chainstate/stacks/miner.rs

```diff
@@ -326,6 +393,218 @@ impl StacksBlockBuilder {
+            None => {
+                // no transactions at this chain tip, so mine an empty block 
+                let block = builder.mine_anchored_block(&mut epoch_tx);
+                builder.epoch_finish(epoch_tx);
+                return Ok(block);
+            }
+        };
+        
+        test_debug!("Begin scanning transaction mempool at {}/{} height={}", &tip_burn_header_hash, &tip_block_hash, tip_height);
+
+        // set to true if we exceed budget, and need to rebuild with known-good transactions.
+        let mut do_rebuild = false;
+
+        loop {
+            let mut available_txs = match MemPoolDB::get_txs_at(mempool.conn(), &tip_burn_header_hash, &tip_block_hash, next_timestamp) {
+                Ok(txs) => txs,
+                Err(e) => {
+                    epoch_tx.rollback_block();
+                    return Err(e.into());
+                }
+            };
+
+            for txinfo in available_txs.drain(..) {
+                if considered.contains(&txinfo.tx.txid()) {
+                    continue;
+                }
+
+                considered.insert(txinfo.tx.txid());
+
+                match builder.try_mine_tx(&mut epoch_tx, &txinfo.tx) {
+                    Ok(_) => {},
+                    Err(Error::BlockTooBigError) => {
+                        // done mining -- our execution budget is exceeded.
+                        // Make the block from the transactions we did manage to get
+                        debug!("Block budget exceeded on tx {}", &txinfo.tx.txid());
+                        do_rebuild = true;
+                    },
+                    Err(e) => {
+                        warn!("Failed to apply tx {}: {:?}", &txinfo.tx.txid(), &e);
+                    }
+                }
+            }
+
+            // next query point
+            match MemPoolDB::get_next_timestamp(mempool.conn(), &tip_burn_header_hash, &tip_block_hash, next_timestamp)? {
+                Some(ts) => {
+                    // still on this tip
+                    next_timestamp = ts;
+                },
+                None => {
+                    // no more transactions at this chain tip.  Walk back to the next-highest
+                    // ancestor of this tip, and see if we can include anything from there.
+                    let next_height = MemPoolDB::get_previous_block_height(mempool.conn(), tip_height)?.unwrap_or(0);
+                    if next_height == 0 && tip_height == 0 {
+                        // we're done -- tried every tx
+                        test_debug!("Done scanning mempool -- at height 0");
+                        break;
+                    }
+
+                    let mut next_tips = MemPoolDB::get_chain_tips_at_height(mempool.conn(), next_height)?;
+                    if next_tips.len() == 0 {
+                        // we're done -- no more chain tips
+                        test_debug!("Done scanning mempool -- no chain tips at height {}", next_height);
```

Comment:
Can these `test_debug`s change to `debug`s?

Reply:
Sure thing.